lib/auth/auth_with_roles.go

```diff
@@ -1281,17 +1284,29 @@ func (a *ServerWithRoles) ListUnifiedResources(ctx context.Context, req *proto.L
 		filter.PredicateExpression = expression
 	}
 
+	// Validate the requested kinds and precheck that the user has read/list
+	// permissions for all requested resources before doing any listing of
+	// resources to conserve resources.
+	requested := req.Kinds
+	if len(req.Kinds) == 0 {
+		requested = defaultUnifiedResourceKinds
+	}
+
 	resourceAccess := &resourceAccess{
 		// Populate kindAccessMap with any access errors the user has for each possible
 		// resource kind. This allows the access check to occur a single time per resource
 		// kind instead of once per matching resource.
-		kindAccessMap: make(map[string]error, len(services.UnifiedResourceKinds)),
+		kindAccessMap: make(map[string]error, len(requested)),
 		// requestableMap is populated with resources that are being returned but can only
 		// be accessed to the user via an access request
 		requestableMap: make(map[string]struct{}),
 	}
 
-	for _, kind := range services.UnifiedResourceKinds {
+	for _, kind := range requested {
+		if _, ok := services.SupportedUnifiedResourceKinds[kind]; !ok {
+			return nil, trace.BadParameter("Unsupported kind %q requested", kind)
+		}
```

Comment:
Are we good with rejecting requests for unsupported kinds? I don't know if the previous way was deliberately lax with checks but it did have the consequence (nice or not) of allowing requests that include unknown types to still go through.